File: backend/apps/admissions/eligibility.py
"""
admissions/eligibility.py — age rule for pre-registration (BACKLOG P1-G1).

School rule (WhatsApp, 2026-08-21): the child must have the required age on
**31 December of the current year**:

    Maternal      2 years
    1° Preescolar 3
    2° Preescolar 4
    3° Preescolar 5
    1° Primaria   6

Higher grades follow the same +1 per grade progression (2° Primaria 7 … 3°
Secundaria 14). The check is strict on the public form; admins can still log
exceptions from the console.
"""
from __future__ import annotations

import re
from datetime import date

from django.utils import timezone

_BASE = {'maternal': 2, 'preescolar': 2, 'primaria': 5, 'secundaria': 11}


def required_age(grade: str) -> int | None:
    """Required age for a grade label like 'Preescolar 1°', '1° Primaria', 'Maternal'."""
    g = (grade or '').lower()
    if 'maternal' in g:
        return _BASE['maternal']
    m = re.search(r'(\d)', g)
    n = int(m.group(1)) if m else None
    for level in ('preescolar', 'primaria', 'secundaria'):
        if level in g:
            return _BASE[level] + (n or 1)
    return None


def age_on_dec31(dob: date, year: int | None = None) -> int:
    year = year or timezone.localdate().year
    ref = date(year, 12, 31)
    return ref.year - dob.year - ((ref.month, ref.day) < (dob.month, dob.day))


def eligibility_error(dob: date, grade: str) -> str | None:
    """None when eligible; otherwise a Spanish message for the form."""
    need = required_age(grade)
    if need is None or dob is None:
        return None
    have = age_on_dec31(dob)
    if have == need:
        return None
    year = timezone.localdate().year
    return (f'Para {grade} el alumno debe tener {need} años cumplidos al 31 de diciembre de {year} '
            f'(tendrá {have}). Elija el grado que corresponde a su edad o contacte a admisiones.')

File: frontend/src/lib/eligibility.ts
/**
 * eligibility.ts — mirror of backend apps/admissions/eligibility.py (BACKLOG P1-G1).
 * Required age on 31 December of the current year per grade.
 */
const BASE: Record<string, number> = { maternal: 2, preescolar: 2, primaria: 5, secundaria: 11 };

export function requiredAge(grade: string): number | null {
  const g = (grade || '').toLowerCase();
  if (g.includes('maternal')) return BASE.maternal;
  const n = Number(/(\d)/.exec(g)?.[1] ?? 1);
  for (const level of ['preescolar', 'primaria', 'secundaria']) {
    if (g.includes(level)) return BASE[level] + n;
  }
  return null;
}

export function ageOnDec31(dobIso: string, year = new Date().getFullYear()): number | null {
  const m = /^(\d{4})-(\d{2})-(\d{2})/.exec(dobIso || '');
  if (!m) return null;
  const y = Number(m[1]); const mo = Number(m[2]); const d = Number(m[3]);
  const before = mo < 12 || (mo === 12 && d <= 31);
  return year - y - (before ? 0 : 1);
}

/** Null when eligible (or unknown); otherwise the Spanish hint shown under the grade. */
export function eligibilityHint(dobIso: string, grade: string): string | null {
  const need = requiredAge(grade);
  const have = ageOnDec31(dobIso);
  if (need === null || have === null) return null;
  if (have === need) return null;
  const year = new Date().getFullYear();
  return `Para ${grade} el alumno debe tener ${need} años al 31 de diciembre de ${year} (tendrá ${have}).`;
}
